pkg/bloomgateway/cache.go

```diff
@@ -0,0 +1,177 @@
+package bloomgateway
+
+import (
+	"context"
+	"flag"
+	"sort"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/chunk/cache/resultscache"
+)
+
+type CacheConfig struct {
+	resultscache.Config `yaml:",inline"`
+	Parallelism         int `yaml:"parallelism"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *CacheConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix(f, "bloom-gateway.cache.")
+	f.IntVar(&cfg.Parallelism, "bloom-gateway.cache.parallelism", 1, "Maximum number fo concurrent requests to gateway server.")
+}
+
+type keyGen struct {
+	Limits
+}
+
+func NewCacheKeyGen(limits Limits) resultscache.KeyGenerator {
+	return keyGen{limits}
+}
+
+func (k keyGen) GenerateCacheKey(ctx context.Context, userID string, r resultscache.Request) string {
```

Comment:
We use `tenant` throughout the bloomgateway and bloomcompactor code:

```suggestion
func (k keyGen) GenerateCacheKey(ctx context.Context, tenant string, r resultscache.Request) string {
```

Reply:
Done